aiida/engine/runners.py

```diff
@@ -263,27 +264,65 @@ def run_get_pk(self, process, *args, **inputs):
         result, node = self._run(process, *args, **inputs)
         return ResultAndPk(result, node.pk)
 
-    def call_on_calculation_finish(self, pk, callback):
-        """
-        Callback to be called when the calculation of the given pk is terminated
+    def call_on_process_finish(self, pk, callback):
+        """Schedule a callback when the process of the given pk is terminated.
 
-        :param pk: the pk of the calculation
-        :param callback: the function to be called upon calculation termination
-        """
-        calculation = load_node(pk=pk)
-        self._poll_calculation(calculation, callback)
+        This method will add a broadcast subscriber that will listen for state changes of the target process to be
+        terminated. As a fail-safe, a polling-mechanism is used to check the state of the process, should the broadcast
+        message be missed by the subscriber, in order to prevent the caller to wait indefinitely.
 
-    def get_calculation_future(self, pk):
+        :param pk: pk of the process
+        :param callback: function to be called upon process termination
         """
-        Get a future for an orm Calculation. The future will have the calculation node
-        as the result when finished.
-
-        :return: A future representing the completion of the calculation node
+        process = load_node(pk=pk)
+        subscriber_identifier = str(uuid.uuid4())
+        resolved = False
+
+        def inline_callback(resolved):
+            """Callback to wrap the actual callback, that will always remove the subscriber that will be registered.
+
+            As soon as the callback is called successfully once, the `resolved` flag is toggled, such that if this
+            inline callback is called a second time, the actual callback is not called again.
+            """
+            if resolved:
+                return
+
+            try:
+                callback()
+            finally:
+                resolved = True
+                try:
+                    self._communicator.remove_broadcast_subscriber(subscriber_identifier)
+                except ValueError:
+                    # The broadcast subscriber has already removed
+                    pass
+
+        broadcast_filter = kiwipy.BroadcastFilter(functools.partial(inline_callback, resolved), sender=pk)
```

Comment:
This, unfortunately, doesn't work.  Python has this sneaky thing where all primitives are passed by value other types by reference meaning that your `resolved` in the callback is actually always a new (`False`) boolean each time the function is invoked.

You can test it with this:
```
res = False

def do(res):
    if res:
        print("Already resolved")
        return
    res = True
    print("Now resolved")



do(res)
do(res)
```


As ugly as it is, this is one solution:
```
res = [False]

def do(res):
    if res[0]:
        print("Already resolved")
        return
    res[0] = True
    print("Now resolved")



do(res)
do(res)
```

But if you don't like it feel free to create anything where an non-primitive is used as a flag.  I think at some point I may have created a special `Flag` class just for this in some of my code.  A `threading.Event` is also a decent choice.